### ISSUE TITLE
Strange bracketed forall loop compiles

### ISSUE THREAD

**Issue by dlongnecke-cray** [chapel-lang/chapel#17814]
While working on square-bracketed forall expressions in the parser
for `compiler/next`, I saw that one of the productions looked
like:

```
| TLSBR expr_ls TRSBR stmt
```

Worried about whether I would need the new interface to support
multiple iterand expressions, I wrote a test program:

```chapel
var i, j = 0;
[i, j] writeln(i, j);
```

This compiles and runs. Looking at the logs, the compiler creates
a domain out of (i, j) and iterates over it, while the loop body
prints the outer variables i, j.

Should this work? If not, is there ever a scenario where the expr
list in the above rule could have more than 1 expression?

If it cannot, should we change `expr_ls` to `expr`?

**Comment by dlongnecke-cray** [comment]
@vasslitvinov I am curious about your thoughts on this.

**Comment by bradcray** [comment]
> should we change expr_ls to expr?

I think this is the right thing to do.  I'd be curious whether anything in the test system breaks with that modification.

**Comment by mppf** [comment]
I'm worried that there might be some array type expression using that grammar production but that just means we need to try it.

**Comment by dlongnecke-cray** [comment]
Checking the old compiler, it looks like this sort of behavior is intentional:

```
| TLSBR expr_ls TRSBR expr %prec TFOR
...
if ($2->argList.length > 1) {
  $$ = buildForallLoopExpr(NULL, new CallExpr("chpl__ensureDomainExpr", $2), $4, NULL, true);
} else {
  $$ = buildForallLoopExpr(NULL, $2->get(1)->remove(), $4, NULL, true);
}
```

**Comment by bradcray** [comment]
Based on that production, my guess is that Michael's guess is right:  That the first case _may_ be used to parse array types.  That is:

`chapel
[i in 1..n] writeln(i);
[1..n] writeln("hi");
...[1..n] int...
```

are all parsed using the same rule, where the last is an array type and the first two are loops.  And this would make sense since we can't generally know syntactically in Chapel whether a given expression is a value expression or a type expression without more context, and full resolution in the most general cases.

But, that doesn't mean that we should let such cases slip through without errors in the value context...

**Comment by dlongnecke-cray** [comment]
I went ahead and made it an error if the `expr_ls` in `TLSBR expr_ls TRSBR expr %prec TFOR` has more than one expression, and got only a single test failure on a `COMM=none` run:

```
  2 [Error matching compiler output for functions/deitz/test_forallexpr3]
```

**Comment by bradcray** [comment]
I don't think that test should work either, according to modern Chapel.  I'd just delete its 3rd and 4th lines and update the .good file accordingly.

**Comment by dlongnecke-cray** [comment]
Ok, I'll do a `COMM=gasnet` run to see if anything else breaks, and then modify the old parser to change `expr_ls` -> `expr`.
